tests/AppBundle/EventListener/RedirectToPreferredLocaleListenerTest.php

```diff
@@ -0,0 +1,99 @@
+
+/**
+ * Unit test for the event listener.
+ *
+ * See http://symfony.com/doc/current/book/testing.html#unit-tests
+ *
+ * Execute the application tests using this command (requires PHPUnit to be installed):
+ *
+ *     $ cd your-symfony-project/
+ *     $ ./vendor/bin/phpunit
+ */
+class RedirectToPreferredLocaleListenerTest extends \PHPUnit_Framework_TestCase
+{
+    private $requestProphecy;
+    private $urlGeneratorProphecy;
+    private $eventProphecy;
+    private $listener;
+
+    public function setUp()
+    {
+        $this->requestProphecy = $this->prophesize(Request::class);
+        $this->urlGeneratorProphecy = $this->prophesize(UrlGeneratorInterface::class);
+        $this->eventProphecy = $this->prophesize(GetResponseEvent::class);
+        $this->listener = new RedirectToPreferredLocaleListener($this->urlGeneratorProphecy->reveal(), 'en|fr|es');
+    }
+
+    public function testOnKernelRequestUnexpectedValueException()
+    {
+        $this->setExpectedException(\UnexpectedValueException::class);
+
+        new RedirectToPreferredLocaleListener($this->urlGeneratorProphecy->reveal(), 'en|fr|es', 'fi');
+    }
+
+    public function testOnKernelRequestWhenIsMasterRequest()
+    {
+        $this->eventProphecy->getRequest()->willReturn($this->requestProphecy->reveal())->shouldBeCalled();
+        $this->eventProphecy->isMasterRequest()->willReturn(false)->shouldBeCalled();
+
+        $this->listener->onKernelRequest($this->eventProphecy->reveal());
+    }
+
+    public function testOnKernelRequestWhenGetPathInfoIsNotDefault()
+    {
+        $this->requestProphecy->getPathInfo()->willReturn('/foo');
+
+        $this->eventProphecy->getRequest()->willReturn($this->requestProphecy->reveal())->shouldBeCalled();
+        $this->eventProphecy->isMasterRequest()->willReturn(false)->shouldBeCalled();
+
+        $this->listener->onKernelRequest($this->eventProphecy->reveal());
+    }
+
+    public function testOnKernelRequestWhenSameHttpHost()
+    {
+        $request = new Request();
+        $request->headers->set('referer', 'http://foo.bar');
+        $request->headers->set('HOST', 'foo.bar');
+
+        $this->eventProphecy->getRequest()->willReturn($request)->shouldBeCalled();
+        $this->eventProphecy->isMasterRequest()->willReturn(true)->shouldBeCalled();
+
+        $this->listener->onKernelRequest($this->eventProphecy->reveal());
+    }
+
+    public function testOnKernelRequestWhenPreferredLanguageNotSame()
+    {
+        $request = new Request();
+        $request->headers->set('referer', 'http://foo.bar');
+        $request->headers->set('HOST', 'bar.fo');
+        $request->headers->set('Accept-Language', 'fr');
+
+        $httpKernelProphecy = $this->prophesize(HttpKernelInterface::class);
+        $event = new GetResponseEvent($httpKernelProphecy->reveal(), $request, HttpKernelInterface::MASTER_REQUEST);
+
+        $this->urlGeneratorProphecy->generate('homepage', ['_locale' => 'fr'])->willReturn('http://foo.sf');
```

Comment:
Shouldn't we test somewhere that the redirected URL is this `http://foo.sf` value?